File: daphne-server/scripts/afe_global.py
"""Independent I283-I288 wire checks; no RPC, hardware access or private output."""

SOURCE = "FPGA:0x80000000,0x9400000C;sequential-read-only"
FIELDS = ("power_state_bit", "reset_asserted", "busy_afe0", "busy_afe12", "busy_afe34", "bias_enabled")


def require(ok, reason="Invalid AFE global observation"):
    if not ok:
        raise RuntimeError(reason)


def check_afe_global(status, high, now_monotonic_ns, *, required=True):
    if not status.HasField("afe_global"):
        require(not required, "Server did not provide AFE global readback")
        return None
    r = status.afe_global
    require(r.quality in (high.MEASUREMENT_UNAVAILABLE, high.MEASUREMENT_GOOD,
                          high.MEASUREMENT_STALE, high.MEASUREMENT_ERROR))
    if r.quality != high.MEASUREMENT_GOOD:
        require(not any(r.HasField(key) for key in FIELDS), "Failed AFE observation retained decoded values")
        require(not required, "AFE global readback is unavailable, stale or failed")
        return {"available": False, "quality": high.MeasurementQuality.Name(r.quality)}
    require(status.success and r.source == SOURCE and r.message and r.maximum_acquisition_ms == 100)
    require(all(r.HasField(key) for key in ("global_control_raw", "bias_enable_raw") + FIELDS))
    require(not r.global_control_raw & ~31 and not r.bias_enable_raw & ~1)
    expected = (bool(r.global_control_raw & 2), bool(r.global_control_raw & 1),
                bool(r.global_control_raw & 4), bool(r.global_control_raw & 8),
                bool(r.global_control_raw & 16), bool(r.bias_enable_raw & 1))
    require(tuple(getattr(r, key) for key in FIELDS) == expected, "AFE flags disagree with raw register bits")
    require(r.identity_bracket_verified and status.HasField("gateware_identity") and status.HasField("fpga_programming"))
    i, p = status.gateware_identity, status.fpga_programming
    require(i.quality == high.MEASUREMENT_GOOD and i.magic == 0x44415048 and
            i.abi in (0x20000, 0x20001, 0x20002) and i.variant in (1, 2) and
            not i.build_id & 0xf0000000 and i.HasField("matches_admitted_profile") and i.matches_admitted_profile)
    require(p.manager_quality == high.MEASUREMENT_GOOD and p.manager_state == "operating" and
            not p.HasField("manager_error_raw") and p.configuration_quality == high.MEASUREMENT_GOOD and
            p.HasField("configuration_status_raw") and not p.configuration_status_raw & 0x28438001 and
            p.configuration_status_raw & 0x78f0 == 0x78f0)
    require(0 < i.acquisition_started_monotonic_ns <= r.acquisition_started_monotonic_ns <=
            r.observed_monotonic_ns <= p.acquisition_started_monotonic_ns <=
            p.manager_observed_monotonic_ns <= p.configuration_observed_monotonic_ns <=
            i.observed_monotonic_ns <= now_monotonic_ns, "AFE observation is outside the admission bracket")
    require(r.observed_monotonic_ns - r.acquisition_started_monotonic_ns <= 100_000_000 and
            now_monotonic_ns - r.observed_monotonic_ns <= 5_000_000_000, "AFE observation is stale")
    return {"available": True, "quality": "MEASUREMENT_GOOD",
            "global_control_raw": f"0x{r.global_control_raw:08x}",
            "bias_enable_raw": f"0x{r.bias_enable_raw:08x}",
            **dict(zip(FIELDS, expected)),
            "acquisition_started_monotonic_ns": r.acquisition_started_monotonic_ns,
            "observed_monotonic_ns": r.observed_monotonic_ns,
            "scope": "sequential FPGA register samples; not bias voltage, physical power, busy history or a health verdict"}

File: daphne-server/srcs/server_controller/afe_global.cpp
#include "server_controller/afe_global.hpp"
#include <stdexcept>

namespace daphne_sc {
bool supports_afe_global(const GatewareIdentity& id) noexcept {
  return id.magic == kGatewareIdentityMagic && supports_gateware_abi(id.abi) &&
      (id.variant == 1 || id.variant == 2) && !(id.build_id & kGatewareBuildIdUpperNibbleMask);
}

void invalidate_afe_global(daphne::AfeGlobalObservation& r,
    daphne::MeasurementQuality quality, const char* reason) {
  r.set_quality(quality); r.set_message(reason);
  r.clear_power_state_bit(); r.clear_reset_asserted();
  r.clear_busy_afe0(); r.clear_busy_afe12(); r.clear_busy_afe34(); r.clear_bias_enabled();
  r.set_identity_bracket_verified(false);
}

bool afe_global_consistent(const daphne::AfeGlobalObservation& r) noexcept {
  if (r.quality() != daphne::MEASUREMENT_GOOD || !r.has_global_control_raw() ||
      !r.has_bias_enable_raw() || (r.global_control_raw() & ~31U) || (r.bias_enable_raw() & ~1U) ||
      !r.has_power_state_bit() || !r.has_reset_asserted() || !r.has_busy_afe0() ||
      !r.has_busy_afe12() || !r.has_busy_afe34() || !r.has_bias_enabled() ||
      r.source() != kAfeGlobalSource || r.maximum_acquisition_ms() != kAfeGlobalMaximumAcquisitionMs ||
      !r.acquisition_started_monotonic_ns() || r.observed_monotonic_ns() < r.acquisition_started_monotonic_ns() ||
      r.observed_monotonic_ns() - r.acquisition_started_monotonic_ns() > uint64_t(kAfeGlobalMaximumAcquisitionMs) * 1000000)
    return false;
  const auto raw = r.global_control_raw();
  return r.reset_asserted() == bool(raw & 1) && r.power_state_bit() == bool(raw & 2) &&
      r.busy_afe0() == bool(raw & 4) && r.busy_afe12() == bool(raw & 8) &&
      r.busy_afe34() == bool(raw & 16) && r.bias_enabled() == bool(r.bias_enable_raw() & 1);
}

daphne::AfeGlobalObservation read_afe_global(Mmio32& afe, Mmio32& bias,
    const GatewareIdentity& admitted, const AfeGlobalClock& clock) {
  daphne::AfeGlobalObservation r;
  r.set_source(kAfeGlobalSource);
  r.set_maximum_acquisition_ms(kAfeGlobalMaximumAcquisitionMs);
  if (!supports_afe_global(admitted)) {
    r.set_message("AFE global readback requires admitted ABI 2.0/2.1/2.2 and a known variant; no registers read");
    return r;
  }
  try {
    r.set_acquisition_started_monotonic_ns(clock());
    if (!r.acquisition_started_monotonic_ns()) throw std::runtime_error("Invalid observation clock");
    r.set_global_control_raw(afe.read32(kAfeGlobalControlAddress));
    r.set_bias_enable_raw(bias.read32(kBiasEnableAddress));
    r.set_observed_monotonic_ns(clock());
    if (r.observed_monotonic_ns() < r.acquisition_started_monotonic_ns() ||
        (r.global_control_raw() & ~31U) || (r.bias_enable_raw() & ~1U))
      throw std::runtime_error("Invalid observation clock or reserved register bits");
    if (r.observed_monotonic_ns() - r.acquisition_started_monotonic_ns() > uint64_t(kAfeGlobalMaximumAcquisitionMs) * 1000000) {
      invalidate_afe_global(r, daphne::MEASUREMENT_STALE, "AFE global reads exceeded their host acquisition budget");
      return r;
    }
    const auto raw = r.global_control_raw();
    r.set_reset_asserted(raw & 1); r.set_power_state_bit(raw & 2);
    r.set_busy_afe0(raw & 4); r.set_busy_afe12(raw & 8); r.set_busy_afe34(raw & 16);
    r.set_bias_enabled(r.bias_enable_raw() & 1);
    r.set_quality(daphne::MEASUREMENT_GOOD);
    r.set_message("Sequential FPGA control-register readback, not physical power/bias voltage, cached SC requests or busy history; outer identity bracket required");
    return r;
  } catch (const std::exception&) {
    invalidate_afe_global(r, daphne::MEASUREMENT_ERROR, "AFE global readback failed; raw evidence is not a qualified observation");
    return r;
  }
}
}
